### PR TITLE
Remove references to Follower.user_id.

### PR DESCRIPTION
**Pull request by ashercodeorg**
This PR finishes making `Followers.user_id` truly optional: at the DB level (#13878), at the Rails validation level (#13945), at the controller level (this PR), and at the Pegasus level (this PR).

After this PR reaches production (safely), further PRs will remove the column from the DB.

### CODE REVIEW

**Comment by ashercodeorg** [comment]
FYI @davidwufer.

File: pegasus/test/test_v2_section_routes.rb [hunk 1]
```diff
@@ -205,15 +205,15 @@ class V2SectionRoutesTest < Minitest::Test
         Dashboard.db.transaction(rollback: :always) do
           before_timestamp = Dashboard.db[:followers].
             where(
-              user_id: FakeDashboard::TEACHER_WITH_DELETED[:id],
+              section_id: SECTION_DELETED_FOLLOWERS[:id],
               student_user_id: FakeDashboard::SELF_STUDENT[:id]
```
Comment (ashercodeorg) [inline pegasus/test/test_v2_section_routes.rb]: That these were being created without `section_id` is somewhat scary.

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -1207,7 +1207,7 @@ def can_edit_password?
   end
 
   def section_for_script(script)
-    followeds.collect(&:section).find {|section| section.script_id == script.id}
+    sections_as_student.find {|section| section.script_id == script.id}
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb:1210]: Tangentially related, this line change could be part of a separate PR.

File: dashboard/app/models/user.rb [hunk 3]
```diff
@@ -1218,14 +1218,7 @@ def terms_version
...
     end
-    # As of August 2016, it may be the case that the `followed` exists but
-    # `followed.user` does not as the result of user deletion. In this case, we
-    # ignore any terms of service versions associated with deleted teacher
-    # accounts.
-    followeds.
-      collect {|followed| followed.user.try(:terms_of_service_version)}.
-      compact.
-      max
+    teachers.map {|teacher| teacher.try(:terms_of_service_version)}.try(:compact).try(:max)
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb]: Tangentially related, this line change could be part of a separate PR.

Comment (aoby) [reply dashboard/app/models/user.rb]: Can `teacher` be nil here? I don't think you need the first `.try`

File: pegasus/helpers/section_api_helpers.rb [hunk 4]
```diff
@@ -105,12 +105,12 @@ def self.fetch_if_allowed_array(ids, dashboard_user_id)
...
-      where(id: params[:id], deleted_at: nil)
+    user_to_update = Dashboard.db[:users].where(id: params[:id], deleted_at: nil)
     return if user_to_update.empty?
-    return if Dashboard.db[:followers].
-      where(student_user_id: params[:id], user_id: dashboard_user_id).
-      where(deleted_at: nil).
+    return if Dashboard.db[:sections].
+      join(:followers, section_id: :sections__id).
+      where(followers__student_user_id: params[:id], sections__user_id: dashboard_user_id).
+      where(followers__deleted_at: nil).
```
Comment (aoby) [inline pegasus/helpers/section_api_helpers.rb]: nit: any reason why the first 2 conditions are in one `where`, and the 3rd condition is in a 2nd `where`? I'd prefer consistency, either all comma-separated args in a single `where` or all separate.

File: dashboard/app/models/user.rb (hunk 3 above)
Comment (ashercodeorg) [reply dashboard/app/models/user.rb]: Done. Good catch.

File: pegasus/helpers/section_api_helpers.rb (hunk 4 above)
Comment (ashercodeorg) [reply pegasus/helpers/section_api_helpers.rb]: Done.
